addons/folo/common/src/main/java/org/commonjava/indy/folo/data/FoloRecordCassandra.java

```diff
@@ -0,0 +1,426 @@
+                DtxTrackingRecord dtxTrackingRecord = new DtxTrackingRecord(entry);
+                dtxTrackingRecord.setLocalUrl(localUrl);
+                trackingMapper.save(dtxTrackingRecord); //  optional Options with TTL, timestamp...
+                return true;
+            }
+
+        } catch (MalformedURLException e) {
+            throw new IndyWorkflowException( "Mailformed URL  for record: {}  path: {}", entry.getTrackingKey(),entry.getPath() );
+        }
+
+
+    }
+
+    @Override
+    public void delete(TrackingKey key) {
+
+        // Do we  want to remove records from DB?
+
+    }
+
+    @Override
+    public void replaceTrackingRecord(TrackedContent record) {
+
+        saveTrackedContentRecords(record);
+
+    }
+
+    @Override
+    public boolean hasRecord(TrackingKey key) {
+        return hasSealedRecord(key) || hasInProgressRecord(key);
+    }
+
+    @Override
+    public boolean hasSealedRecord(TrackingKey key) {
+
+        BoundStatement bind = getTrackingRecordsByTrackingKey.bind(key);
+        ResultSet execute = session.execute(bind);
+        Row one = execute.one();
+        if(one != null) {
+            Boolean sealed = one.getBool("sealed");
+            if(sealed) {
+                return true;
+            } else {
+                return false;
+            }
+        }else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean hasInProgressRecord(TrackingKey key) {
+        BoundStatement bind = getTrackingRecordsByTrackingKey.bind(key);
+        ResultSet execute = session.execute(bind);
+        Row one = execute.one();
+        if(one != null) {
+            Boolean sealed = one.getBool("sealed");
+            if(!sealed) {
+                return true;
+            } else {
+                return false;
+            }
+        }else {
+            return false;
+        }
+    }
+
+    @Override
+    public TrackedContent get(TrackingKey key) {
+        List<DtxTrackingRecord> trackingRecords =  getDtxTrackingRecordsFromDb(key);
+        return transformDtxTrackingRecordToTrackingContent(key,trackingRecords);
+    }
+
+    @Override
+    public TrackedContent seal(TrackingKey trackingKey) {
+        List<DtxTrackingRecord> trackingRecords =  getDtxTrackingRecordsFromDb(trackingKey);
+        DtxTrackingRecord recordCheck = trackingRecords.get(0);
+        if(recordCheck.getState()) {
+            logger.debug( "Tracking record: {} already sealed! Returning sealed record.", trackingKey );
+            return transformDtxTrackingRecordToTrackingContent(trackingKey,trackingRecords);
+        }
+        logger.debug( "Sealing record for: {}", trackingKey );
+        for(DtxTrackingRecord record : trackingRecords) {
+            record.setState(true);
+            trackingMapper.save(record);
+        }
+        return transformDtxTrackingRecordToTrackingContent(trackingKey,trackingRecords);
+    }
+
+    @Override
+    public Set<TrackingKey> getInProgressTrackingKey() {
```

Comment:
We need to do a small refactor here. We use this in the FoloAdminController to determine which records cache to query...so FoloAdminController is over-optimized to a system where we have two cache types storing these records.

Instead, we should push the parameter (sealed / unsealed) down to the FoloRecordCache class / interface level, and move that logic out of the FoloAdminController. Then, in this implementation we can do something more intelligent with filtering at the database level, and get more performance. That should allow us to collapse `getInProgressTrackingKey()` and `getSealedTrackingKey()` into a single method like `getTrackingKeys(FoloConstants.TRACKING_TYPE trackingType)` and push that trackingType parameter directly into a Cassandra query parameter.

Reply:
I agree, i wanna to do that when. i was taking out interface. from FoloRecordCache but i. didn't want to do to much changes that's way i just leave all methods in interface.

Reply:
> We use this in the FoloAdminController to determine which records cache to query...so FoloAdminController is over-optimized to a system where we have two cache types storing these records.

I started consolidating. them in. one method but i. notice. something:
We. could also have "all" where in. Ctrl. we are putting "in_progress" and "sealed" in set and based on that branching in Ctrl is. acquiring both key sets. So i'm not sure how. much performance will be gaining. because we still need to get all tracking keys if they. are. asked from. user side.